consensus/istanbul/backend/handler.go

```diff
@@ -286,3 +301,150 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
+	isValidatorCh := make(chan bool)
+
+	sendHandshake := func() {
+		var validatorProofMsg *istanbul.Message
+		var err error
+		peerIsValidator := peer.PurposeIsSet(p2p.ValidatorPurpose)
+		if peerIsValidator {
+			// validatorProofMsg may be nil
+			validatorProofMsg, err = sb.retrieveVersionedEnodeMsg()
+			if err != nil {
+				errCh <- err
+				return
+			}
+		}
+		// Even if we decide not to identify ourselves,
+		// send an empty message to complete the handshake
+		if validatorProofMsg == nil {
+			validatorProofMsg = &istanbul.Message{}
+		}
+		msgBytes, err := validatorProofMsg.Payload()
+		if err != nil {
+			errCh <- err
+			return
+		}
+		err = peer.Send(istanbulValidatorProofMsg, msgBytes)
+		if err != nil {
+			errCh <- err
+		}
+		isValidatorCh <- peerIsValidator
+	}
+	readHandshake := func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		isValidatorCh <- isValidator
+	}
+
+	// Only the initating peer sends the message
+	if peer.Inbound() {
+		go readHandshake()
+	} else {
+		go sendHandshake()
+	}
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	select {
+	case err := <-errCh:
+		return false, err
+	case <-timeout.C:
+		return false, p2p.DiscReadTimeout
+	case isValidator := <-isValidatorCh:
+		return isValidator, nil
+	}
+}
+
+// readValidatorProofMessage reads a validator proof message as a part of the
+// istanbul handshake. Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorProofMessage(peer consensus.Peer) (bool, error) {
+	logger := sb.logger.New("func", "readValidatorProofMessage")
+	peerMsg, err := peer.ReadMsg()
+	if err != nil {
+		return false, err
+	}
+	if peerMsg.Code != istanbulValidatorProofMsg {
+		logger.Warn("Read incorrect message code", "code", peerMsg.Code)
+		return false, errors.New("Incorrect message code")
+	}
+
+	var payload []byte
+	if err := peerMsg.Decode(&payload); err != nil {
+		return false, err
+	}
+
+	var msg istanbul.Message
+	err = msg.FromPayload(payload, sb.verifyValidatorProofMessage)
+	if err != nil {
+		return false, err
+	}
+	// If the Signature is empty, the peer has decided not to reveal its info
+	if len(msg.Signature) == 0 {
+		return false, nil
+	}
+
+	var versionedEnode versionedEnode
+	err = rlp.DecodeBytes(msg.Msg, &versionedEnode)
+	if err != nil {
+		return false, err
+	}
+
+	node, err := enode.ParseV4(versionedEnode.EnodeURL)
+	if err != nil {
+		return false, err
+	}
+
+	// Ensure the node in the versionedEnode matches the peer node
+	if node.ID() != peer.Node().ID() {
+		logger.Warn("Peer provided incorrect node ID in versionedEnode", "versionedEnode enode url", versionedEnode.EnodeURL, "peer enode url", peer.Node().URLv4())
+		return false, errors.New("Incorrect node in versionedEnode")
+	}
+
+	// Check if the peer is within the registered/elected valset
+	regAndActiveVals, err := sb.retrieveCachedRegisteredAndElectedValidators()
```

Comment:
I just realized that this lets a validator who is registered and not active bypass the max peer check. Right now only the active validators are meant to be a part of the overlay network, but this is meant to change to the "announce set"

@kevjue I know you're working on the announce set, does that involve including the entire announce set in the overlay network or just limiting announce messages to the announce set and keeping peer connections amongst those that are active?